### DISCUSSION TITLE
Export shader

### DISCUSSION THREAD

**Discussion by zeeyang** [curv3d/curv#114] [Q&A] [answered]
I'm looking for a way to compile a curv file and export it as a glsl shader, maybe also a companion descriptor file for the parametrics, so I can render it in another app. @doug-moen if this doesn't exist can you point me to the code path? Maybe I can help implement it. Thanks!

**Comment by doug-moen** [comment]
Your starting point is to export a *.gpu file, like this:
```
$ curv -o thing.gpu thing.curv
```
This feature is not designed for your exact use case, because a `.gpu` file is a Curv source file, and your app doesn't have the ability to read Curv source code. You could try writing an ad hoc parser to extract the GLSL from a `.gpu` file, or we could add a new feature to Curv that is more useable for your purpose. I will think about this some more.

**Comment by doug-moen** [comment, marked as answer]
The function that writes the shape compiler output as a `*.gpu` file is located at
```
libcurv/viewed_shape.cc   Viewed_Shape::write_curv()
```
In that same source file, there is also a function called `write_json()`, which isn't currently being used. That might be a solution to your problem, since whatever programming language your app is written in, you will have a JSON parser library available.

The code for exporting a `*.gpu` file, which calls `Viewed_Shape::write_curv()`, is located in
```
curv/export.cc   export_gpu()
```
So my first idea is that you could add the ability to export a `*.jcpu` file containing JSON code.

**Comment by doug-moen** [comment]
I've got more ideas, but you should describe the requirements for your application in more detail before I respond.

**Reply by zeeyang** [reply]
Thanks @doug-moen. I'll give `write_json` a try.

The application I have in mind is an iOS app with a library of parametric curv models for my 3D printer, so my kids can just tap and print. e.g. the fat bottom mug example. The basic requirements are,
- load and render curv shader using swift
- parse curv parametric fields and provide controls in swift
- make a server side round trip to curv runtime to generate the stl (in the future bundle curv in the app)
- also in the future mutiple shader language output (e.g. metal, vulkan, webgpu)

**Comment by doug-moen** [comment]
Thanks. I want to say something about long term support. This willl work in the current release (0.4) and the next release (0.5). However, in some future release, I hope to add support for images, voxel grids and triangle meshes, as Curv values. These can be imported from a file or constructed by a Curv program. At that point, the `*.gpu` file format would need to contain images, voxel grids and triangle meshes, which could be large, and at some point JSON encoding no longer makes sense. So `*.jgpu` support will eventually be dropped.

Also, Curv will be rewritten from an OpenGL app to a WebGPU app (there is native support for WebGPU on Linux, Windows, MacOS, Android, iOS). Hopefully that happens this year; we'll see. The MacOS/iOS version of WebGPU uses Metal as the back end. WebGPU accepts SPIR-V or WGSL input for shaders, and compiles to MSL. My current thought is to continue generating GLSL, as I currently do, then convert to SPIR-V using a library, and then that is transpiled to MSL by the WebGPU library. (Apple will remove OpenGL support at some point, so I can't rely on it existing in the future.) At this point, the GPU programming model will also change. Shapes may no longer be exclusively encoded as fragment shaders. You may not care about any of this, but if you want to update your app to the latest version of Curv a year or two from now, there will be significant changes of a sort that I haven't fully planned out.
